lib/pure/asyncdispatch.nim

```diff
@@ -850,6 +850,241 @@ when defined(windows) or defined(nimdoc):
+    )
+
+    let res = WSARecvFrom(socket.SocketHandle, addr dataBuf, 1,
+                          addr bytesReceived, addr lowFlags,
+                          saddr, cast[ptr cint](saddrLen),
+                          cast[POVERLAPPED](ol), nil)
+    if res == -1:
+      let err = osLastError()
+      if err.int32 != ERROR_IO_PENDING:
+        GC_unref(ol)
+        retFuture.fail(newException(OSError, osErrorMsg(err)))
+    else:
+      # Request completed immediately.
+      if bytesReceived != 0:
+        assert bytesReceived <= size
+        retFuture.complete(bytesReceived)
+      else:
+        if hasOverlappedIoCompleted(cast[POVERLAPPED](ol)):
+          retFuture.complete(bytesReceived)
+    return retFuture
+
+  template recvFromTemplate(socket: AsyncFD, recvSockType, retFut) =
+    var dataBuf: TWSABuf
+    # maximum size of UDP datagram data is 65536
+    let size = 65536.ULONG
+
+    dataBuf.buf = cast[cstring](alloc0(size))
+    dataBuf.len = size
+
+    var bytesReceived = 0.Dword
+    var lowFlags = 0.Dword
+    var saddr = recvSockType()
+    var saddrLen = sizeof(recvSockType).SockLen
+
+    var ol = PCustomOverlapped()
+    GC_ref(ol)
+    ol.data = CompletionData(fd: socket, cb:
+      proc (fd: AsyncFD, bytesCount: Dword, errcode: OSErrorCode) =
+        if not retFut.finished:
+          if errcode == OSErrorCode(-1):
+            if bytesCount == 0:
+              retFut.complete((data: "", saddr: saddr))
+            else:
+              assert(bytesCount <= size)
+              var data = newString(bytesCount)
+              copyMem(addr(data[0]), addr(dataBuf.buf[0]), bytesCount)
+              retFut.complete((data: data, saddr: saddr))
+          else:
+            retFut.fail(newException(OSError, osErrorMsg(errcode)))
+        if dataBuf.buf != nil:
+          dealloc(dataBuf.buf)
+          dataBuf.buf = nil
+    )
+
+    let res = WSARecvFrom(socket.SocketHandle, addr dataBuf, 1,
+                          addr bytesReceived, addr lowFlags,
+                          cast[ptr SockAddr](addr saddr),
+                          cast[ptr cint](addr saddrLen),
+                          cast[POVERLAPPED](ol), nil)
+    if res == -1:
+      let err = osLastError()
+      if err.int32 != ERROR_IO_PENDING:
+        if dataBuf.buf != nil:
+          dealloc dataBuf.buf
+          dataBuf.buf = nil
+        GC_unref(ol)
+        retFut.fail(newException(OSError, osErrorMsg(err)))
+    else:
+      # Request completed immediately.
+      if bytesReceived != 0:
+        assert(bytesReceived <= size)
+        var data = newString(bytesReceived)
+        copyMem(addr data[0], addr dataBuf.buf[0], bytesReceived)
+        retFut.complete((data: data, saddr: saddr))
+      else:
+        if hasOverlappedIoCompleted(cast[POVERLAPPED](ol)):
+          retFut.complete((data: "", saddr: saddr))
+
+  proc recvFrom*(socket: AsyncFD, flags = {SocketFlag.SafeDisconn}):
+       Future[tuple[data: string, saddr: Sockaddr_in]] =
```

Comment:
I dont think its possible, because recvFrom handle buffer deallocation in callback in case of failure.